tests: Add multicast-pim-bsm-topo1 suite

kuldeepkash opened the pull request:
1. Added 15 testcases to verify PIM BSM functionality. Here we have used Scapy
    to send raw packet, generated using Cisco and FRR. Raw packets are kept in
    JSON file and sent tests on-demand in script.
2. Added framework support for multicast-pim-bsm-topo1 suite automation
3. Execution time is ~20 min.

Signed-off-by: Kuldeep Kashyap <kashyapk@vmware.com>

kuldeepkash commented:
CI:rerun

donaldsharp commented:
please rebase and force push

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/78e392099d114420529b782e6dcc1358/raw/d757d8585d9a030818d53ef9079e2c2747941ddc/cr_7218_1602568588.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index d83f946c4..d7f2c1502 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2780,7 +2780,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2797,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

kuldeepkash commented:
Hi @mwinter-osr, We need "scapy==2.4.2" package to be installed on VMs, where topotests is running. We are using scapy to send BSM raw packets(which are already generated and kept in JSON file). 

Please suggest and help to proceed futher.

donaldsharp commented:
please update the doc/developer/topotest.rst file with this new library addition

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/577a32a45f376d42d1627e93a90f382f/raw/d757d8585d9a030818d53ef9079e2c2747941ddc/cr_7218_1602761253.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index d83f946c4..d7f2c1502 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2780,7 +2780,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2797,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

mwinter-osr commented on tests/topotests/lib/common_config.py:
```diff
@@ -31,14 +31,16 @@
 import traceback
 import socket
 import ipaddress
+import ipaddr
```
ipaddr was superseeded by the ipaddress module. With the work on getting topotests to run on Python3 (with limited ipaddr support), is there a specific need to use ipaddr instead of ipaddress?
If not, then please use ipaddress instead

kuldeepkash replied:
I have replaced ipaddr with ipaddress. Thanks @mwinter-osr

mwinter-osr reviewed (changes requested):
Still need to add scapy, but in the meantime, please address the ipaddr issue

mwinter-osr commented:
CI:rerun

mwinter-osr commented:
scapy is now added to testbeds. Please note the error in the test `test_mcast_pim_bsmp_01` `test_BSR_higher_prefer_ip_p0` on the latest CI testrun

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/67f8427a83ab7ada3fa35a74026735a3/raw/d757d8585d9a030818d53ef9079e2c2747941ddc/cr_7218_1603788031.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index d83f946c4..d7f2c1502 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2780,7 +2780,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2797,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

donaldsharp reviewed (changes requested):
1) *why* do I keep seeing commits removing verify_fib_routes?
2) *why* is *bgp* being started
3) This commit is 6k lines, *it* must be broken up more into small commits that people can understand what is being done.  I see lots of work changing stuff around in the tests, which is fine but it's all jumbled around and lots of stuff is not related to each other.

donaldsharp commented on tests/topotests/lib/topojson.py:
```diff
@@ -151,24 +144,24 @@ def link_sort(x):
-                    destRouter_link_json = topo["routers"][destRouter]["links"][
-                        curRouterLink
-                    ]
+
```
Whitespace changes are fine, but need to be *pulled* out into their own commits.

kuldeepkash replied:
Removed whitespace from this commit. We will take care in another commit.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/d48fa2685cb3759e5ea5cdae19cdffed/raw/d757d8585d9a030818d53ef9079e2c2747941ddc/cr_7218_1604395518.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index d83f946c4..d7f2c1502 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2780,7 +2780,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2797,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

kuldeepkash commented:
> 1. _why_ do I keep seeing commits removing verify_fib_routes?
> 2. _why_ is _bgp_ being started
> 3. This commit is 6k lines, _it_ must be broken up more into small commits that people can understand what is being done.  I see lots of work changing stuff around in the tests, which is fine but it's all jumbled around and lots of stuff is not related to each other.

Hi @donaldsharp, Please find the comment below:
1. I have deleted API verify_fib_routes(), because by mistake it is duplicated in libraries, so I have deleted one.
2. Update code so only required daemons as per feature test would be started.
3. I have split the code into 2 different commits.

Please look into it and provide your valuable feedback/suggestions.

kuldeepkash commented on tests/topotests/lib/common_config.py:
```diff
@@ -3182,259 +3213,6 @@ def verify_fib_routes(tgen, addr_type, dut, input_dict, next_hop=None):
     return True
 
 
-@retry(attempts=5, wait=2, return_is_str=True, initial_wait=2)
```
Hi Donald, 

This API is somehow duplicated in code, so deleting one.

kuldeepkash commented on tests/topotests/lib/common_config.py:
```diff
@@ -828,13 +844,17 @@ def topo_daemons(tgen, topo):
     router_list = tgen.routers()
     ROUTER_LIST = sorted(
-        router_list.keys(), key=lambda x: int(re_search("[0-9]+", x).group(0))
+        router_list.keys(), key=lambda x: int(re_search("\d+", x).group(0))
```
Thanks Donald, I have reverted these changes, as not needed here.

kuldeepkash commented on tests/topotests/lib/common_config.py:
```diff
@@ -715,6 +720,10 @@ def start_topology(tgen, daemon=None):
             logger.info("Logging platform related details: \n %s \n", linux_ver)
 
+        # Listing pip packages
+        pip_packages = router.run("pip freeze")
```
Thanks Donald, I have reverted these changes, as not needed here.

kuldeepkash commented on tests/topotests/lib/common_config.py:
```diff
@@ -700,7 +705,7 @@ def start_topology(tgen, daemon=None):
     router_list = tgen.routers()
     ROUTER_LIST = sorted(
-        router_list.keys(), key=lambda x: int(re_search("[0-9]+", x).group(0))
+        router_list.keys(), key=lambda x: int(re_search("\d+", x).group(0))
```
Thanks Donald, I have reverted these changes, as not needed here.

kuldeepkash commented on tests/topotests/lib/common_config.py:
```diff
@@ -638,7 +643,7 @@ def get_frr_ipv6_linklocal(tgen, router, intf=None, vrf=None):
 
             # Interface ip
-            m1 = re_search("inet6 (fe80[:a-fA-F0-9]+[/0-9]+)", line)
+            m1 = re_search("inet6 (fe80[:a-fA-F0-9]+[\/0-9]+)", line)
```
Thanks Donald, I have reverted these changes, as not needed here.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/b006ca132fe28171bdf44446db308a8a/raw/d757d8585d9a030818d53ef9079e2c2747941ddc/cr_7218_1604403834.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index d83f946c4..d7f2c1502 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2780,7 +2780,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2797,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/5b8c1ccfe3324297a774e925a154c530/raw/d757d8585d9a030818d53ef9079e2c2747941ddc/cr_7218_1604417066.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index d83f946c4..d7f2c1502 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2780,7 +2780,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2797,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

kuldeepkash commented:
CI:rerun

kuldeepkash commented:
CI:rerun

kuldeepkash commented:
CI:rerun

kuldeepkash commented:
@donaldsharp, Could you please look into the changes and help in reviewing and merging this PR. Thank you.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/5b63b896fb3cdef7557822e221a5b85a/raw/46482ab5245fff25fdd6874717929d8c7571cdb9/cr_7218_1605159178.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index e42992e29..9ac749869 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2559,7 +2559,7 @@ def verify_rib(
     tag=None,
     metric=None,
     fib=None,
-    count_only=False
+    count_only=False,
 ):
     """
     Data will be read from input_dict or input JSON file, API will generate
@@ -2751,8 +2751,10 @@ def verify_rib(
                                             "Nexthops are missing for "
                                             "route {} in RIB of router {}: "
                                             "expected {}, found {}\n".format(
-                                                st_rt, dut, len(next_hop),
-                                                len(found_hops)
+                                                st_rt,
+                                                dut,
+                                                len(next_hop),
+                                                len(found_hops),
                                             )
                                         )
                                         return errormsg
@@ -2799,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2816,7 +2822,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/20fdf95658bfa9e5e8f6b3e6ce17b184/raw/46482ab5245fff25fdd6874717929d8c7571cdb9/cr_7218_1605168023.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index e42992e29..9ac749869 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2559,7 +2559,7 @@ def verify_rib(
     tag=None,
     metric=None,
     fib=None,
-    count_only=False
+    count_only=False,
 ):
     """
     Data will be read from input_dict or input JSON file, API will generate
@@ -2751,8 +2751,10 @@ def verify_rib(
                                             "Nexthops are missing for "
                                             "route {} in RIB of router {}: "
                                             "expected {}, found {}\n".format(
-                                                st_rt, dut, len(next_hop),
-                                                len(found_hops)
+                                                st_rt,
+                                                dut,
+                                                len(next_hop),
+                                                len(found_hops),
                                             )
                                         )
                                         return errormsg
@@ -2799,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2816,7 +2822,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/aaab88950ebe12c264e627b0a2a105f7/raw/46482ab5245fff25fdd6874717929d8c7571cdb9/cr_7218_1605188753.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index e42992e29..9ac749869 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2559,7 +2559,7 @@ def verify_rib(
     tag=None,
     metric=None,
     fib=None,
-    count_only=False
+    count_only=False,
 ):
     """
     Data will be read from input_dict or input JSON file, API will generate
@@ -2751,8 +2751,10 @@ def verify_rib(
                                             "Nexthops are missing for "
                                             "route {} in RIB of router {}: "
                                             "expected {}, found {}\n".format(
-                                                st_rt, dut, len(next_hop),
-                                                len(found_hops)
+                                                st_rt,
+                                                dut,
+                                                len(next_hop),
+                                                len(found_hops),
                                             )
                                         )
                                         return errormsg
@@ -2799,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2816,7 +2822,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

mwinter-osr reviewed (approved):
Tests verified and run stable un multiple runs. Go to go from my side

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/16fc8e6cf11a392f8e5104209524ee2c/raw/ca34a37db6ac0957a04c8a05dc9d9bbb0a1c41bf/cr_7218_1605686038.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index 9c104db3c..145a67bf8 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -2559,7 +2559,7 @@ def verify_rib(
     tag=None,
     metric=None,
     fib=None,
-    count_only=False
+    count_only=False,
 ):
     """
     Data will be read from input_dict or input JSON file, API will generate
@@ -2751,8 +2751,10 @@ def verify_rib(
                                             "Nexthops are missing for "
                                             "route {} in RIB of router {}: "
                                             "expected {}, found {}\n".format(
-                                                st_rt, dut, len(next_hop),
-                                                len(found_hops)
+                                                st_rt,
+                                                dut,
+                                                len(next_hop),
+                                                len(found_hops),
                                             )
                                         )
                                         return errormsg
@@ -2799,7 +2801,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2816,7 +2822,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -2944,6 +2954,7 @@ def verify_rib(
     logger.debug("Exiting lib API: {}".format(sys._getframe().f_code.co_name))
     return True
 
+
 @retry(attempts=6, wait=2, return_is_str=True)
 def verify_fib_routes(tgen, addr_type, dut, input_dict, next_hop=None):
     """

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

kuldeepkash commented:
CI:rerun

kuldeepkash commented:
CI:rerun

kuldeepkash commented:
CI:rerun

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/e4f904ff1c571041be3d990b9c5b9859/raw/975019665dbdaff55317763560ace5889655d30f/cr_7218_1610430801.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index 6e2e23cff..95371c6aa 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -744,9 +744,7 @@ def start_topology(tgen, daemon=None):
         )
 
         # Loading empty bgpd.conf file to router, to start the bgp deamon
-        router.load_config(
-            TopoRouter.RD_BGP, "{}/{}/bgpd.conf".format(TMPDIR, rname)
-        )
+        router.load_config(TopoRouter.RD_BGP, "{}/{}/bgpd.conf".format(TMPDIR, rname))
 
         if daemon and "ospfd" in daemon:
             # Loading empty ospf.conf file to router, to start the bgp daemon
@@ -3067,7 +3065,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: tag value {}"
                                         " is not matched for"
-                                        " route {} in RIB \n".format(dut, _tag, st_rt,)
+                                        " route {} in RIB \n".format(
+                                            dut,
+                                            _tag,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
@@ -3084,7 +3086,11 @@ def verify_rib(
                                     errormsg = (
                                         "[DUT: {}]: metric value "
                                         "{} is not matched for "
-                                        "route {} in RIB \n".format(dut, metric, st_rt,)
+                                        "route {} in RIB \n".format(
+                                            dut,
+                                            metric,
+                                            st_rt,
+                                        )
                                     )
                                     return errormsg
 
diff --git a/tests/topotests/lib/pim.py b/tests/topotests/lib/pim.py
index 45aa3534e..f1f80d3d0 100644
--- a/tests/topotests/lib/pim.py
+++ b/tests/topotests/lib/pim.py
@@ -1537,7 +1537,11 @@ def verify_pim_interface(tgen, topo, dut):
                     errormsg = (
                         "[DUT %s]: PIM interface: %s "
                         "PIM interface ip: %s, is not present "
-                        % (dut, pim_interface, pim_intf_ip,)
+                        % (
+                            dut,
+                            pim_interface,
+                            pim_intf_ip,
+                        )
                     )
                     return errormsg
 
diff --git a/tests/topotests/lib/topojson.py b/tests/topotests/lib/topojson.py
index 431a2a3f2..bb07feb2c 100644
--- a/tests/topotests/lib/topojson.py
+++ b/tests/topotests/lib/topojson.py
@@ -42,9 +42,7 @@ from lib.common_config import (
     create_bgp_community_lists,
     create_vrf_cfg,
 )
-from lib.pim import (
-    create_pim_config, create_igmp_config
-)
+from lib.pim import create_pim_config, create_igmp_config
 from lib.bgp import create_router_bgp
 from lib.ospf import create_router_ospf
 
@@ -100,32 +98,42 @@ def build_topo_from_json(tgen, topo):
     while listRouters != []:
         curRouter = listRouters.pop(0)
         # Physical Interfaces
-        if 'links' in topo['routers'][curRouter]:
-            for destRouterLink, data in sorted(topo['routers'][curRouter]['links']. \
-                                               iteritems()):
-                currRouter_lo_json = \
-                    topo['routers'][curRouter]['links'][destRouterLink]
+        if "links" in topo["routers"][curRouter]:
+            for destRouterLink, data in sorted(
+                topo["routers"][curRouter]["links"].iteritems()
+            ):
+                currRouter_lo_json = topo["routers"][curRouter]["links"][destRouterLink]
                 # Loopback interfaces
-                if 'type' in data and data['type'] == 'loopback':
+                if "type" in data and data["type"] == "loopback":
                     router_count += 1
-                    currRouter_lo_json['interface'] = destRouterLink
-
-                    if 'ipv4' in currRouter_lo_json and \
-                            currRouter_lo_json['ipv4'] == 'auto':
-                        currRouter_lo_json['ipv4'] = '{}{}.{}/{}'. \
-                            format(topo['lo_prefix']['ipv4'], router_count, \
-                                   number_to_column(curRouter), topo['lo_prefix']['v4mask'])
-
-                    if 'ipv6' in currRouter_lo_json and \
-                            currRouter_lo_json['ipv6'] == 'auto':
-                        currRouter_lo_json['ipv6'] = '{}{}:{}/{}'. \
-                            format(topo['lo_prefix']['ipv6'], router_count, \
-                                   number_to_column(curRouter), topo['lo_prefix']['v6mask'])
+                    currRouter_lo_json["interface"] = destRouterLink
+
+                    if (
+                        "ipv4" in currRouter_lo_json
+                        and currRouter_lo_json["ipv4"] == "auto"
+                    ):
+                        currRouter_lo_json["ipv4"] = "{}{}.{}/{}".format(
+                            topo["lo_prefix"]["ipv4"],
+                            router_count,
+                            number_to_column(curRouter),
+                            topo["lo_prefix"]["v4mask"],
+                        )
+
+                    if (
+                        "ipv6" in currRouter_lo_json
+                        and currRouter_lo_json["ipv6"] == "auto"
+                    ):
+                        currRouter_lo_json["ipv6"] = "{}{}:{}/{}".format(
+                            topo["lo_prefix"]["ipv6"],
+                            router_count,
+                            number_to_column(curRouter),
+                            topo["lo_prefix"]["v6mask"],
+                        )
 
                 if "dst_node" in data:
-                    destRouter = data['dst_node']
-                    if 'dst_link' in data:
-                        curRouterLink = data['dst_link']
+                    destRouter = data["dst_node"]
+                    if "dst_link" in data:
+                        curRouterLink = data["dst_link"]
                     else:
                         curRouterLink = destRouterLink
                 elif "-" in destRouterLink:
@@ -177,7 +185,7 @@ def build_topo_from_json(tgen, topo):
                         ],
                         topo["routers"][destRouter]["links"][curRouterLink][
                             "interface"
-                        ]
+                        ],
                     )
 
                     # IPv4

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

donaldsharp commented on tests/topotests/lib/common_config.py:
```diff
@@ -743,8 +743,10 @@ def start_topology(tgen, daemon=None):
 
-        # Loading empty bgpd.conf file to router, to start the bgp daemon
-        router.load_config(TopoRouter.RD_BGP, "{}/{}/bgpd.conf".format(TMPDIR, rname))
+        # Loading empty bgpd.conf file to router, to start the bgp deamon
```
Why this whitespace change?

kuldeepkash replied:
It is resolved, changes are merged via different PR.

donaldsharp commented on tests/topotests/lib/pim.py:
```diff
@@ -466,76 +466,6 @@ def find_rp_details(tgen, topo):
     return rp_details
 
 
-def configure_pim_force_expire(tgen, topo, input_dict, build=False):
```
Why is this being deleted?

kuldeepkash replied:
It is resolved, changes are merged via different PR.

donaldsharp commented on tests/topotests/lib/topojson.py:
```diff
@@ -99,36 +100,35 @@ def build_topo_from_json(tgen, topo):
-                        )
-
-                if "-" in destRouterLink:
+                    currRouter_lo_json['interface'] = destRouterLink
```
Haven't we gone over this before?  Why are we mixing whitespace changes with other code changes.  It makes it extremely hard to review and is plain annoying to wade through.  Please take some time and go through your commits and break them *up*

kuldeepkash replied:
It is resolved, changes are merged via different PR.

donaldsharp reviewed (changes requested):
Please cleanup this review

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/1271de612c186c968c2a868eff63003b/raw/c58d4780bb3ffb8a2da1b8d486e527d670ed2523/cr_7218_1610993616.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/lib/common_config.py b/tests/topotests/lib/common_config.py
index 175d660d1..c8f54d975 100644
--- a/tests/topotests/lib/common_config.py
+++ b/tests/topotests/lib/common_config.py
@@ -4395,4 +4395,3 @@ def verify_ip_nht(tgen, input_dict):
 
     logger.debug("Exiting lib API: verify_ip_nht()")
     return False
-
diff --git a/tests/topotests/lib/pim.py b/tests/topotests/lib/pim.py
index 6bb132651..96b553a6c 100644
--- a/tests/topotests/lib/pim.py
+++ b/tests/topotests/lib/pim.py
@@ -1607,7 +1607,11 @@ def verify_pim_interface(tgen, topo, dut):
                     errormsg = (
                         "[DUT %s]: PIM interface: %s "
                         "PIM interface ip: %s, is not present "
-                        % (dut, pim_interface, pim_intf_ip,)
+                        % (
+                            dut,
+                            pim_interface,
+                            pim_intf_ip,
+                        )
                     )
                     return errormsg
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

kuldeepkash commented:
CI:rerun

kuldeepkash commented:
Hi @donaldsharp , Could you please have a look of this PR. It is updated as per your suggestions, given earlier. Please check once and help is in merging it. Thank you.
